text/0000-local-loop-bindings.md

```diff
@@ -0,0 +1,218 @@
+- Feature Name: local_loop_bindings
+- Start Date: 2018-12-25
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This will allow an extended syntax for `loop` to accept local variables, which change for each iteration to simplify more complicated loop constructs and avoid mutable state.
+
+
+# Motivation
+[motivation]: #motivation
+
+The new syntax is inspired by `loop` in the upcoming release of the [scopes programming language](scopes.rocks).
+This way it's possible use different values for each iteration without the need of mutable variables defined outside of the loop.
+
+The variables will be defined after the loop keyword, so they will only be accessible in the scope of the loop, not afterwards. They will not be mutable by default, so it can be ensured, that the variables only change once per iteration.
+
+Especially since loops can return values, it's not necessary at all to mutate state inside a loop in some cases.
+
+This is a more functional programming style, which may also allow more optimizations like storing the loop arguments in registers instead of allocating storage for mutable variables.
```

Comment:
I'm not sure, how to explain. I won't do this yet

Reply:
Alright; take your time. :)

Reply:
FWIW, mutable local variables are regularly converted to registers by LLVM.  I wouldn't expect any optimization differences here, especially since the new syntax would disappear by the time we get to MIR anyway.